File: functions/prompts/guidelines/theminjoo.js
'use strict';

const { db } = require('../../utils/firebaseAdmin');

/**
 * 🏛️ 더불어민주당 공식 당론 및 어조 가이드 중앙 저장소 (DB 기반)
 * 
 * 역할:
 * 1. Firestore `party_stances` 컬렉션에서 주제와 관련된 당론 조회
 * 2. 매칭되는 당론이 있을 경우 프롬프트에 주입할 가이드 객체 반환
 */

// 컬렉션 이름
const COLLECTION_NAME = 'party_stances';

// 캐싱 (메모리 내) - 콜드 스타트 시에만 DB 조회, 이후 재사용 (선택 사항)
// 현재는 실시간성이 중요하므로 캐싱하지 않거나 짧게 유지
const stanceCache = null;
const lastCacheTime = 0;
const CACHE_TTL = 60 * 1000; // 1분

/**
 * 주제(topic)와 관련된 당론을 조회합니다.
 * @param {string} topic - 사용자가 입력한 주제
 * @returns {Promise<Object|null>} 당론 객체 또는 null
 */
async function getPartyStance(topic) {
  if (!topic) return null;

  try {
    // 1. 모든 활성화된 당론 가져오기 (문서 수가 적을 것으로 가정하고 전체 로드 후 필터링)
    // 문서 수가 많아지면 쿼리 최적화 필요
    const snapshot = await db.collection(COLLECTION_NAME).where('isActive', '==', true).get();

    if (snapshot.empty) {
      return null;
    }

    let matchedStance = null;
    let maxMatchedKeywords = 0;

    // 2. 키워드 매칭 로직
    // 가장 많은 키워드가 매칭된 당론을 선택 (단순 포함 여부)
    snapshot.docs.forEach(doc => {
      const data = doc.data();
      const keywords = data.keywords || [];

      let matchCount = 0;
      keywords.forEach(keyword => {
        if (topic.includes(keyword)) {
          matchCount++;
        }
      });

      // 매칭된 키워드가 있고, 기존보다 더 많이 매칭되었거나 (우선순위 고려 가능)
      if (matchCount > 0 && matchCount > maxMatchedKeywords) {
        maxMatchedKeywords = matchCount;
        matchedStance = { id: doc.id, ...data };
      }
    });

    if (matchedStance) {
      console.log(`🏛️ [PartyStance] 당론 매칭 성공: "${matchedStance.title}" (키워드 매칭: ${maxMatchedKeywords}개)`);
      return formatStanceForPrompt(matchedStance);
    }

    return null;

  } catch (error) {
    console.error('❌ [PartyStance] 당론 조회 실패:', error);
    return null;
  }
}

/**
 * 당론 데이터를 프롬프트용 텍스트로 변환
 */
function formatStanceForPrompt(stanceData) {
  const forbiddenPhrases = (stanceData.forbidden_phrases || []).join(', ');

  return `
╔═══════════════════════════════════════════════════════════════╗
║  🏛️ [CRITICAL] 더불어민주당 공식 당론 가이드 (자동 적용)      ║
╚═══════════════════════════════════════════════════════════════╝

**관련 이슈**: ${stanceData.title}

1. **[핵심 입장 - Stance]**:
   "${stanceData.stance}"

2. **[필수 논리 구조 - Logic Guide]**:
   ${stanceData.logic_guide}

3. **[절대 금지 표현]**:
   - 금지: ${forbiddenPhrases}
   ${stanceData.additional_instructions ? `\n4. **[추가 지침]**:\n   ${stanceData.additional_instructions}` : ''}
`;
}

module.exports = {
  getPartyStance
};

File: functions/python/agents/common/theminjoo.py
import logging
from firebase_admin import firestore

COLLECTION_NAME = 'party_stances'

logger = logging.getLogger("theminjoo")

def get_db():
    return firestore.client()

def format_stance_for_prompt(stance_data):
    """Formats the stance data into a string for the LLM prompt."""
    forbidden = stance_data.get('forbidden_phrases', [])
    forbidden_str = ", ".join(forbidden)

    additional = ""
    if stance_data.get('additional_instructions'):
        additional = f"\n4. **[추가 지침]**:\n   {stance_data['additional_instructions']}"

    return f"""
╔═══════════════════════════════════════════════════════════════╗
║  🏛️ [CRITICAL] 더불어민주당 공식 당론 가이드 (자동 적용)      ║
╚═══════════════════════════════════════════════════════════════╝

**관련 이슈**: {stance_data.get('title')}

1. **[핵심 입장 - Stance]**:
   "{stance_data.get('stance')}"

2. **[필수 논리 구조 - Logic Guide]**:
   {stance_data.get('logic_guide')}

3. **[절대 금지 표현]**:
   - 금지: {forbidden_str}
   {additional}
"""

def get_party_stance(topic: str):
    """
    Retrieves the most relevant party stance based on the topic.
    Returns the formatted string or None.
    """
    if not topic:
        return None

    try:
        db = get_db()
        # Fetch all active stances
        docs = db.collection(COLLECTION_NAME).where('isActive', '==', True).stream()
        
        matched_stance = None
        max_matched_keywords = 0

        for doc in docs:
            data = doc.to_dict()
            keywords = data.get('keywords', [])
            
            match_count = 0
            for keyword in keywords:
                if keyword in topic:
                    match_count += 1
            
            if match_count > 0 and match_count > max_matched_keywords:
                max_matched_keywords = match_count
                data['id'] = doc.id
                matched_stance = data

        if matched_stance:
            logger.info(f"🏛️ [PartyStance] Found match: '{matched_stance.get('title')}' ({max_matched_keywords} keywords)")
            return format_stance_for_prompt(matched_stance)
        
        return None

    except Exception as e:
        logger.warning(f"❌ [PartyStance] Failed to fetch stances: {e}. Using fallback.")
        return format_stance_for_prompt({
            'title': '더불어민주당 일반 강령 (Fallback)',
            'stance': '민생 우선, 정의로운 대한민국',
            'logic_guide': '- 서민 경제 회복 강조\n- 민주주의 가치 수호',
            'forbidden_phrases': ['독재 타협', '부자 감세'],
            'additional_instructions': '이 내용은 DB 연결 실패 시 제공되는 기본값입니다.'
        })
